### PR TITLE
Allow directly passing explicit font paths.

### PR DESCRIPTION
**Pull request by anntzer**
This allows one to use `text(..., font=Path("/path/to/font.ttf"))`
rather than the more cumbersome
`fontproperties=FontProperties(fname="/path/to/font.ttf")`.

Note that we do not allow `font="/path/to/font.ttf"` and instead only
support Path objects because we also already support `font="DejaVu Sans"`
and I would rather not guess whether we have been passed a path or a
family (if there is really demand for that we can always add it later
anyways).

See https://github.com/matplotlib/matplotlib/issues/10249.

Needs whatsnew, but just posting this here for discussion.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/axis.py [hunk 1]
```diff
@@ -164,14 +163,10 @@ def __init__(self, axes, loc, label=None,
             GRIDLINE_INTERPOLATION_STEPS
         self.label1 = mtext.Text(
             np.nan, np.nan,
-            fontproperties=font_manager.FontProperties(size=labelsize),
```
Comment (anntzer) [inline lib/matplotlib/axis.py]: there are a few cleanups like this which are technically independent and could be extracted to a separate PR if needed.

File: lib/mpl_toolkits/axisartist/axis_artist.py [hunk 2]
```diff
@@ -942,21 +941,19 @@ def _init_ticks(self, axis_name, **kwargs):
         else:
             size = rcParams['ytick.labelsize']
 
-        fontprops = font_manager.FontProperties(size=size)
```
Comment (anntzer) [inline lib/mpl_toolkits/axisartist/axis_artist.py]: this one too

File: lib/matplotlib/textpath.py [hunk 3]
```diff
@@ -400,6 +400,8 @@ def __init__(self, xy, s, size=None, prop=None,
 
         if prop is None:
             prop = FontProperties()
+        else:
+            prop = FontProperties._from_any(prop)
```
Comment (Kojoley) [inline lib/matplotlib/textpath.py]: Did you forget to copy result, like you have done above? Maybe `_from_any` should return copy always?

File: lib/matplotlib/text.py [hunk 4]
```diff
@@ -1222,9 +1222,7 @@ def set_fontproperties(self, fp):
         ----------
         fp : `.font_manager.FontProperties`
         """
-        if isinstance(fp, str):
-            fp = FontProperties(fp)
-        self._fontproperties = fp.copy()
+        self._fontproperties = FontProperties._from_any(fp).copy()
```
Comment (Kojoley) [inline lib/matplotlib/text.py]: Maybe `_from_any` should return copy always?

File: lib/matplotlib/textpath.py (hunk 3 above)
Comment (anntzer) [reply lib/matplotlib/textpath.py]: For now I decided to keep the previous (probably bad) semantics where someone could e.g. pass in a FontProperties instance and, depending on how the thing was written, changes to the *original* instance may or may not affect the downstream artist.  I agree always making a copy is likely to result in more consistent behavior, but let's not change that here.

File: lib/matplotlib/axis.py (hunk 1 above)
Comment (timhoffm) [reply lib/matplotlib/axis.py]: I always like smaller PRs better :smile:.

File: lib/matplotlib/text.py (hunk 4 above)
Comment (anntzer) [reply lib/matplotlib/text.py]: See below

File: lib/matplotlib/axis.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/axis.py]: Sure, it was split into https://github.com/matplotlib/matplotlib/pull/16246.

File: examples/text_labels_and_annotations/font_file.py [hunk 5]
```diff
@@ -4,8 +4,8 @@
 ===================================
 
 Although it is usually not a good idea to explicitly point to a single ttf file
-for a font instance, you can do so using the `.font_manager.FontProperties`
-*fname* argument.
+for a font instance, you can do so by passing a `pathlib.Path` instance as the
+*font* parameter.
```
Comment (timhoffm) [inline examples/text_labels_and_annotations/font_file.py]: Should mention that strings are intentionally not supported, but you can simply wrap them in a `Path` if necessary.

Comment (anntzer) [reply examples/text_labels_and_annotations/font_file.py]: done

**Comment by anntzer** [comment]
Added some docs mentioning this to Text.set_fontproperties, FontProperties, and a whatsnew.
